scylla/src/transport/topology.rs

```diff
@@ -869,6 +870,181 @@ mod control_connection {
             self.conn.execute_iter(prepared_statement, values).await
         }
     }
+
+    #[cfg(test)]
+    mod tests {
+        use std::net::SocketAddr;
+        use std::sync::Arc;
+        use std::time::Duration;
+
+        use scylla_proxy::{
+            Condition, Node, Proxy, Reaction as _, RequestFrame, RequestOpcode, RequestReaction,
+            RequestRule, ShardAwareness,
+        };
+        use tokio::sync::mpsc;
+
+        use crate::test_utils::setup_tracing;
+        use crate::transport::connection::open_connection;
+        use crate::transport::connection::tests::resolve_hostname;
+        use crate::transport::node::ResolvedContactPoint;
+        use crate::transport::topology::{ConnectionConfig, UntranslatedEndpoint};
+
+        use super::ControlConnection;
+
+        /// Tests that ControlConnection enforces the provided custom timeout
+        /// iff ScyllaDB is the target node (else ignores the custom timeout).
+        #[cfg(not(scylla_cloud_tests))]
+        #[tokio::test]
+        #[ntest::timeout(2000)]
+        async fn test_custom_timeouts() {
+            setup_tracing();
+
+            let proxy_addr = SocketAddr::new(scylla_proxy::get_exclusive_local_address(), 9042);
+            let uri = std::env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());
+            let node_addr: SocketAddr = resolve_hostname(&uri).await;
+
```

Comment:
❓ This test requires a running Scylla / C* cluster, right? Is it possible to avoid this?

Reply:
It is possible to avoid this by using the proxy in the _dry mode_. We could mock the whole CQL handshake (`OPTIONS -> SUPPORTED -> STARTUP -> READY -> REGISTER -> READY`) and then intercept `QUERY`s and `PREPARE`s as before. An additional gain would be to test multiple scenarios at ones, i.e. sharded and non-sharded endpoints, independently of the actual cluster being deployed for tests.

Reply:
I've made the test use the dry mode to avoid dependency on a running cluster.